### PR TITLE
fix:[DPMMA-2818] fixes sort reports table

### LINKED ISSUE #14102: Sorting by table headers in view report not Working
### Mautic Series

5.1.x series

### Mautic installed version

5.1.0

### Way of installing

I downloaded a release from https://www.mautic.org/mautic-releases

### PHP version

8.2

### What browsers are you seeing the problem on?

Chrome

### What happened?

Sorting by clicking on the table headers in the report data table is not functioning as expected. When clicking a column header to sort the data (either ascending or descending), the table does not reorder the data accordingly. This issue persists across multiple columns, regardless of the data type (e.g., numeric, date, or text fields).

image

### How can we reproduce this issue?

Step 1: Navigate to the report data table.
Step 2: Click on any column header to sort the data by that column.
Step 3: Observe that the data remains unsorted or does not change as expected.

Expected Behavior: Clicking on the column header should sort the data in the table based on the selected column, either in ascending or descending order.

Current Behavior: Clicking the column header does not sort the data.

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

### PR DESCRIPTION
**Pull request by AlanWierzchonCA**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 
| New feature/enhancement? (use the a.x branch)      | 🔴
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 

## Description

 Fix issue: Sorting by table headers in view report not Working - https://github.com/mautic/mautic/issues/14102

image

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

2: Navigate to the report view with data table.
3: Click on any column header to sort the data by that column.
4: Verify that the data remains sorted as expected.

### CODE REVIEW

File: app/bundles/ReportBundle/Resources/views/Report/_details_report_content.html.twig [hunk 1]
```diff
@@ -24,7 +24,11 @@
                             {% for key in columnOrder %}
                                 {% if columns[key] is defined %}
                                     {# order by alias if exists, if not then by column name #}
-                                    {%- set orderBy = columns[key].alias|default((key starts with 'channel.') ? key|replace({'.': '_'}) : key) -%}
+                                    {%- set orderBy = (columns[key].alias is defined and columns[key].alias starts with 'channel.')
+                                        ? columns[key].alias|replace({'.': '-'})
+                                        : key|replace({'.': '-'})
+                                    -%}
```
Comment (patrykgruszka) [inline app/bundles/ReportBundle/Resources/views/Report/_details_report_content.html.twig]: I can't find a reason why the `.` character is replaced here. It works fine with the dot.

**Review by patrykgruszka — CHANGES_REQUESTED** [review]
The report sorting works fine after this fix.

Two suggestions:
1. Remove the dot character replacement, as it seems unnecessary.
2. Add a functional test to cover this case. Example: https://github.com/mautic/mautic/blob/5.x/app/bundles/PointBundle/Tests/Functional/ReportSubscriberFunctionalTest.php

File: app/bundles/ReportBundle/Resources/views/Report/_details_report_content.html.twig (hunk 1 above)
Comment (AlanWierzchonCA) [reply app/bundles/ReportBundle/Resources/views/Report/_details_report_content.html.twig]: I changed this.

File: app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php [hunk 2]
```diff
@@ -449,6 +488,11 @@ public function testDescriptionIsNotEscaped(): void
         $this->assertStringContainsString('<small><b>This is allowed HTML</b></small>', $clientResponseContent);
     }
 
+    private function domTableToArray(Crawler $crawler): array
```
Comment (patrykgruszka) [inline app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php:520]:
```suggestion
    /**
     * @return array<int,array<int,mixed>>
     */
    private function domTableToArray(Crawler $crawler): array
```
Fix for PHPSTAN

File: app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php [hunk 3]
```diff
@@ -53,6 +54,44 @@ public function testMostVisitedPagesReport(): void
...
 
+    public function testReportTableOrderColumn(): void
+    {
+        $page  = $this->createPage('test page 1', 15);
+        $page2 = $this->createPage('test page 2', 9);
+        $page3 = $this->createPage('test page 3', 30);
+
+        $this->createHit($page);
+        $this->createHit($page2);
+        $this->createHit($page3);
```
Comment (patrykgruszka) [inline app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php:65]:
```suggestion
```
This is probably not needed, it creates only one hit per page

File: app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php [hunk 4]
```diff
@@ -53,6 +54,44 @@ public function testMostVisitedPagesReport(): void
...
+
+        $report = $this->createReport('Report Most Visited Pages', 'page.hits', []);
+        $report->setColumns(['p.title', 'p.hits']);
+        $report->setGraphs(['mautic.email.table.most.emails.clicks']);
```
Comment (patrykgruszka) [inline app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php]:
```suggestion
```
We are not testing any graphs here, so we can remove this

**Review by patrykgruszka — CHANGES_REQUESTED** [review]
I've added few suggestions to the test and we need to fix the PHPSTAN

File: app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php (hunk 3 above)
Comment (AlanWierzchonCA) [reply app/bundles/ReportBundle/Tests/Controller/ReportControllerFunctionalTest.php:65]: Without this there are no report results.

**Review by patrykgruszka — APPROVED** [review]
Thanks, looks fine :+1:

**Review by andersonjeccel — APPROVED** [review]
Finally works, thanks for that


https://github.com/user-attachments/assets/726b15c3-f72e-4758-8c62-ac0ec74425e6

just found a bit weird that you have to click twice for a filter to start working, but it might not be related to the PR...

**Pull request merged**